// maverick.js/dist/dev/chunks/maverick-jhjCtVsR.js
/**
 * Converts a camelCase string to kebab-case.
 *
 * @example `myProperty -> my-property`
 */
function camelToKebabCase(str) {
  return str.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
}

/**
 * Converts a camelCase string to Title Case.
 *
 * @example `myProperty -> Title Case`
 */
function camelToTitleCase(str) {
  return uppercaseFirstChar(str.replace(/([A-Z])/g, ' $1'));
}

/**
 * Converts a kebab-case string to camelCase.
 *
 * @example `my-property -> myProperty`
 */
function kebabToCamelCase(str) {
  return str.replace(/-./g, (x) => x[1].toUpperCase());
}

/**
 * Converts a kebab-case string to PascalCase.
 *
 * @example `myProperty -> MyProperty`
 */
function kebabToPascalCase(str) {
  return kebabToTitleCase(str).replace(/\s/g, '');
}

/**
 * Converts a kebab-case string to Title Case.
 *
 * @example `myProperty -> My Property`
 */
function kebabToTitleCase(str) {
  return uppercaseFirstChar(str.replace(/-./g, (x) => ' ' + x[1].toUpperCase()));
}

function uppercaseFirstChar(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

function lowercaseFirstLetter(str) {
  return str.charAt(0).toLowerCase() + str.slice(1);
}

const trailingSemicolon = /;\s*$/;
function trimTrailingSemicolon(text) {
  return text.replace(trailingSemicolon, '');
}

export { camelToKebabCase, camelToTitleCase, kebabToCamelCase, kebabToPascalCase, kebabToTitleCase, lowercaseFirstLetter, trimTrailingSemicolon, uppercaseFirstChar };
